### PR TITLE
UCP: preparations for the tag matching offload

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please review

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1547/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3466/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3472/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1553/ for details.

**Comment by Di0gen** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3600/ for details (Mellanox internal link).

File: src/ucp/core/ucp_worker.c [hunk 1]
```diff
@@ -299,11 +310,24 @@ static ucs_status_t ucp_worker_add_iface(ucp_worker_h worker,
         }
     }
 
+    if (ucp_worker_is_tl_tag_offload(worker, tl_id)) {
+        if (ucs_queue_is_empty(&context->tm.offload_ifaces)) {
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: unite to common function with 'add' flow

File: src/ucp/tag/tag_recv.c [hunk 2]
```diff
@@ -45,6 +45,7 @@ ucp_tag_search_unexp(ucp_worker_h worker, void *buffer, size_t buffer_size,
             ucp_tag_log_match(recv_tag, rdesc->length - rdesc->hdr_len, req, tag,
                               tag_mask, req->recv.state.offset, "unexpected");
             ucs_queue_del_iter(&context->tm.unexpected, iter);
+
```
Comment (yosefe) [inline src/ucp/tag/tag_recv.c]: pls remove adding space lines

File: src/ucp/tag/tag_send.c [hunk 3]
```diff
@@ -126,14 +125,14 @@ static void ucp_tag_req_start_generic(ucp_request_t *req, size_t count,
...
-    if (length >= rndv_am_thresh) {
-        /* rendezvous */
-        ucp_tag_send_start_rndv(req);
-        UCS_PROFILE_REQUEST_EVENT(req, "start_rndv", req->send.length);
-    } else if (length <= config->am.max_bcopy - proto->only_hdr_size) {
+    if (length <= config->tag.eager.max_bcopy - proto->only_hdr_size) {
         /* bcopy single */
         req->send.uct.func = proto->bcopy_single;
         UCS_PROFILE_REQUEST_EVENT(req, "start_gen_bcopy_single", req->send.length);
+    } else if ((length >= rndv_am_thresh) || config->tag.offload.enabled) {
```
Comment (yosefe) [inline src/ucp/tag/tag_send.c]: can we set rndv_am_thresh to lower value if offload is enabled?

File: src/ucp/wireup/address.h [hunk 4]
```diff
@@ -24,6 +24,7 @@ enum {
          UCT_IFACE_FLAG_PUT_BCOPY |
          UCT_IFACE_FLAG_GET_BCOPY |
          UCT_IFACE_FLAG_GET_ZCOPY |
+         UCT_IFACE_FLAG_TAG_EAGER_BCOPY |
```
Comment (yosefe) [inline src/ucp/wireup/address.h]: need to add RNDV

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1711/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3648/ for details (Mellanox internal link).

File: src/ucp/tag/tag_send.c [hunk 5]
```diff
@@ -40,8 +41,11 @@ static ucs_status_t ucp_tag_req_start(ucp_request_t *req, size_t count,
         req->send.state.dt.iov.iovcnt_offset = 0;
         req->send.state.dt.iov.iov_offset    = 0;
         req->send.state.dt.iov.iovcnt        = count;
-        flag_iov_single                      = (count <= config->tag.eager.max_iov) ||
-                                                config->tag.offload.enabled;
+        flag_iov_single                      = (count <= config->tag.eager.max_iov);
+        if ((!flag_iov_single) && ucp_ep_is_tag_offload_enabled(config)) {
```
Comment (yosefe) [inline src/ucp/tag/tag_send.c]: no need for extra ( )

File: src/ucp/tag/rndv.c [hunk 6]
```diff
@@ -72,8 +72,7 @@ static size_t ucp_tag_rndv_rts_pack(void *dest, void *arg)
     rndv_rts_hdr->sreq.reqptr      = (uintptr_t)sreq;
     rndv_rts_hdr->sreq.sender_uuid = sreq->send.ep->worker->uuid;
     rndv_rts_hdr->size             = sreq->send.length;
-    if (UCP_DT_IS_CONTIG(sreq->send.datatype) ||
-        UCP_DT_IS_IOV(sreq->send.datatype)) {
+    if (UCP_DT_IS_CONTIG(sreq->send.datatype)) {
```
Comment (yosefe) [inline src/ucp/tag/rndv.c]: is it reverting the previous PR?

File: src/ucp/tag/tag_send.c [hunk 7]
```diff
@@ -40,8 +41,11 @@ static ucs_status_t ucp_tag_req_start(ucp_request_t *req, size_t count,
...
+            rndv_thresh = length; /* make sure SW RNDV will be used */
```
Comment (yosefe) [inline src/ucp/tag/tag_send.c]: maybe need another flag - force_sw_rdnv ?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1715/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3654/ for details (Mellanox internal link).

**Pull request merged**